library/std/src/sys_common/rwlock.rs

```diff
@@ -1,88 +1,81 @@
     pub const fn new() -> RWLock {
-        RWLock(imp::RWLock::new())
+        RWLock { inner: imp::RWLock::new(), _pinned: PhantomPinned }
     }
 
     /// Acquires shared access to the underlying lock, blocking the current
     /// thread to do so.
-    ///
-    /// Behavior is undefined if the rwlock has been moved between this and any
-    /// previous method call.
     #[inline]
-    pub unsafe fn read(&self) {
-        self.0.read()
+    pub fn read(self: Pin<&Self>) {
+        unsafe { self.inner.read() }
     }
 
     /// Attempts to acquire shared access to this lock, returning whether it
     /// succeeded or not.
     ///
     /// This function does not block the current thread.
-    ///
-    /// Behavior is undefined if the rwlock has been moved between this and any
-    /// previous method call.
     #[inline]
-    pub unsafe fn try_read(&self) -> bool {
-        self.0.try_read()
+    pub fn try_read(self: Pin<&Self>) -> bool {
+        unsafe { self.inner.try_read() }
     }
 
     /// Acquires write access to the underlying lock, blocking the current thread
     /// to do so.
-    ///
-    /// Behavior is undefined if the rwlock has been moved between this and any
-    /// previous method call.
     #[inline]
-    pub unsafe fn write(&self) {
-        self.0.write()
+    pub fn write(self: Pin<&Self>) {
+        unsafe { self.inner.write() }
     }
 
     /// Attempts to acquire exclusive access to this lock, returning whether it
     /// succeeded or not.
     ///
     /// This function does not block the current thread.
-    ///
-    /// Behavior is undefined if the rwlock has been moved between this and any
-    /// previous method call.
     #[inline]
-    pub unsafe fn try_write(&self) -> bool {
-        self.0.try_write()
+    pub fn try_write(self: Pin<&Self>) -> bool {
+        unsafe { self.inner.try_write() }
     }
 
     /// Unlocks previously acquired shared access to this lock.
     ///
     /// Behavior is undefined if the current thread does not have shared access.
     #[inline]
-    pub unsafe fn read_unlock(&self) {
-        self.0.read_unlock()
+    pub unsafe fn read_unlock(self: Pin<&Self>) {
+        self.inner.read_unlock()
     }
 
     /// Unlocks previously acquired exclusive access to this lock.
     ///
     /// Behavior is undefined if the current thread does not currently have
     /// exclusive access.
     #[inline]
-    pub unsafe fn write_unlock(&self) {
-        self.0.write_unlock()
+    pub unsafe fn write_unlock(self: Pin<&Self>) {
+        self.inner.write_unlock()
     }
+}
 
-    /// Destroys OS-related resources with this RWLock.
-    ///
-    /// Behavior is undefined if there are any currently active users of this
-    /// lock.
+impl Drop for RWLock {
     #[inline]
-    pub unsafe fn destroy(&self) {
-        self.0.destroy()
+    fn drop(&mut self) {
+        // SAFETY: The rwlock wasn't moved since using any of its
+        // functions, because they all require a Pin.
```

Comment:
Isn't this changing the safety preconditions? Previously there was something about "active users", which is no longer mentioned.

This seems related to https://github.com/rust-lang/rust/issues/85434. The "no active users" condition is not actually upheld by the users of this code. So it makes sense to remove this comment here, but IMO then there should be a comment one layer down -- because there `pthread_mutex_destroy` is being called and the safety precondition does not ensure all that needs to be ensured. (Basically, this moves the "active users" comment instead of removing it entirely, and acknowledges that we have a gap in our safety reasoning here.)